src/common/os_auto_flush_linux.c

```diff
@@ -0,0 +1,190 @@
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * os_auto_flush_linux.c -- Linux abstraction layer for auto flush detection
+ */
+
+#define _GNU_SOURCE
+
+#include <inttypes.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <string.h>
+#include "out.h"
+#include "os.h"
+#include "fs.h"
+#include "os_auto_flush.h"
+
+#define BUS_DEVICE_PATH "/sys/bus/nd/devices"
+#define PERSISTENCE_DOMAIN "persistence_domain"
+#define DOMAIN_VALUE_LEN 32
+
+/*
+ * os_auto_flush -- check if platform supports auto flush for all regions
+ *
+ * Traverse "/sys/bus/nd/devices" path to find all the nvdimm regions,
+ * then for each region checks if "persistence_domain" file exists and
+ * contains "cpu_cache" string.
+ * If for any region "persistence_domain" entry does not exists, or its
+ * context is not as expected, assume eADR is not available on this platform.
+ */
+int
+os_auto_flush(void)
+{
+	LOG(15, NULL);
+
+	char *device_path;
+	char region_path[PATH_MAX];
+	char domain_path[PATH_MAX];
+	char domain_value[DOMAIN_VALUE_LEN];
+	int domain_fd = -1;
+	int CPU_cache = 0;
+	struct fs *reg = NULL;
+
+	device_path = BUS_DEVICE_PATH;
+
+	os_stat_t sdev;
+	if (os_stat(device_path, &sdev) != 0 ||
+		S_ISDIR(sdev.st_mode) == 0) {
+		LOG(3, "eADR not supported");
+		return CPU_cache;
+	}
+
+	struct fs *dev = fs_new(device_path);
+	if (dev == NULL) {
+		ERR("!fs_new: \"%s\"", device_path);
+		return -1;
+	}
+
+	struct fs_entry *dev_entry;
+	struct fs_entry *reg_entry;
+
+	while ((dev_entry = fs_read(dev)) !=  NULL) {
+		/*
+		 * skip if not a symlink,
+		 * does not contain region in name,
+		 * or depth different than 1
+		 */
+		if ((dev_entry->type != FS_ENTRY_SYMLINK) ||
+				!strstr(dev_entry->name, "region") ||
+				dev_entry->level != 1)
+			continue;
+
+		LOG(15, "start %s traversing", dev_entry->name);
+		int ret = snprintf(region_path, PATH_MAX, "%s/%s",
```

Comment:
isn't this dev_entry->path ?

Instead of nesting fs instances, you could expand the interface to include children traversal. And regardless of what you do, I'd prefer to split this into a couple of functions.